### PR TITLE
[Serve] Thread through ReplicaConfig instead of each individual replica config

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
This will make it easier to add new replica configs.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/deployment.py [hunk 1]
```diff
@@ -163,57 +153,38 @@ def __init__(
...
         self._is_driver_deployment = is_driver_deployment
         self._docs_path = docs_path
 
+    @property
+    def deployment_config(self) -> DeploymentConfig:
+        return self._deployment_config
+
+    @property
+    def replica_config(self) -> ReplicaConfig:
+        return self._replica_config
```
Comment (edoakes) [inline python/ray/serve/deployment.py:186]: `Deployment` is a public API and `ReplicaConfig` / `DeploymentConfig` currently aren't, so let's make these accessors private please

File: python/ray/serve/tests/test_replica_placement_group.py [hunk 2]
```diff
@@ -236,14 +236,11 @@ def leaked_pg_cleaned_up():
...
 
-    @serve.deployment(
-        placement_group_bundles=[{"CPU": 0.1}],
-    )
     class Infeasible:
         pass
 
     with pytest.raises(ValueError):
-        serve.run(Infeasible.bind())
+        serve.deployment(placement_group_bundles=[{"CPU": 0.1}])(Infeasible)
```
Comment (edoakes) [inline python/ray/serve/tests/test_replica_placement_group.py:243]: pls also test `.options()` path

**Review by edoakes — COMMENTED** [review]
One comment otherwise LGTM, this will be a big improvement!

**Review by sihanwang41 — APPROVED** [review]
applaud, love it.

**Pull request merged**